File: src/vfs/policy.ts
import { parentOf } from './path-utils.js';
import { vfsError } from './errors.js';

export interface VfsResourcePolicy {
  maxFileBytes?: number;
  maxTotalBytes?: number;
  maxDirectoryDepth?: number;
  maxEntriesPerDirectory?: number;
  maxOutputArtifactBytes?: number;
}

export interface VfsSnapshot {
  files: Map<string, number>;
  dirs: Set<string>;
}

export interface VfsPolicyOptions {
  resourcePolicy?: VfsResourcePolicy;
}

export type ResolvedVfsResourcePolicy = Readonly<VfsResourcePolicy>;

const EMPTY_RESOURCE_POLICY: ResolvedVfsResourcePolicy = Object.freeze({});

export function resolveVfsResourcePolicy(policy?: VfsResourcePolicy): ResolvedVfsResourcePolicy {
  if (!policy) {
    return EMPTY_RESOURCE_POLICY;
  }

  const resolved: VfsResourcePolicy = {};
  validateLimit('maxFileBytes', policy.maxFileBytes);
  validateLimit('maxTotalBytes', policy.maxTotalBytes);
  validateLimit('maxDirectoryDepth', policy.maxDirectoryDepth);
  validateLimit('maxEntriesPerDirectory', policy.maxEntriesPerDirectory);
  validateLimit('maxOutputArtifactBytes', policy.maxOutputArtifactBytes);

  if (policy.maxFileBytes !== undefined) {
    resolved.maxFileBytes = policy.maxFileBytes;
  }
  if (policy.maxTotalBytes !== undefined) {
    resolved.maxTotalBytes = policy.maxTotalBytes;
  }
  if (policy.maxDirectoryDepth !== undefined) {
    resolved.maxDirectoryDepth = policy.maxDirectoryDepth;
  }
  if (policy.maxEntriesPerDirectory !== undefined) {
    resolved.maxEntriesPerDirectory = policy.maxEntriesPerDirectory;
  }
  if (policy.maxOutputArtifactBytes !== undefined) {
    resolved.maxOutputArtifactBytes = policy.maxOutputArtifactBytes;
  }

  return Object.freeze(resolved);
}

export function hasResourcePolicy(policy: ResolvedVfsResourcePolicy): boolean {
  return Object.keys(policy).length > 0;
}

export function cloneSnapshot(snapshot: VfsSnapshot): VfsSnapshot {
  return {
    files: new Map(snapshot.files),
    dirs: new Set(snapshot.dirs),
  };
}

export function createEmptySnapshot(): VfsSnapshot {
  return {
    files: new Map(),
    dirs: new Set(['/']),
  };
}

export function applyMkdirToSnapshot(snapshot: VfsSnapshot, path: string, parents: boolean): void {
  if (path === '/') {
    return;
  }

  if (parents) {
    ensureSnapshotParents(snapshot, path);
    return;
  }

  snapshot.dirs.add(path);
}

export function applyWriteToSnapshot(
  snapshot: VfsSnapshot,
  path: string,
  size: number,
  makeParents: boolean,
): void {
  if (makeParents) {
    ensureSnapshotParents(snapshot, parentOf(path));
  }
  snapshot.files.set(path, size);
}

export function applyAppendToSnapshot(
  snapshot: VfsSnapshot,
  path: string,
  sizeDelta: number,
  makeParents: boolean,
): void {
  if (makeParents) {
    ensureSnapshotParents(snapshot, parentOf(path));
  }
  snapshot.files.set(path, (snapshot.files.get(path) ?? 0) + sizeDelta);
}

export function applyDeleteToSnapshot(snapshot: VfsSnapshot, path: string): void {
  if (snapshot.files.has(path)) {
    snapshot.files.delete(path);
    return;
  }

  const prefix = `${path}/`;
  for (const filePath of [...snapshot.files.keys()]) {
    if (filePath.startsWith(prefix)) {
      snapshot.files.delete(filePath);
    }
  }
  for (const dirPath of [...snapshot.dirs]) {
    if (dirPath === path || dirPath.startsWith(prefix)) {
      snapshot.dirs.delete(dirPath);
    }
  }
  snapshot.dirs.add('/');
}

export function applyCopyToSnapshot(snapshot: VfsSnapshot, src: string, dst: string): void {
  if (snapshot.files.has(src)) {
    ensureSnapshotParents(snapshot, parentOf(dst));
    snapshot.files.set(dst, snapshot.files.get(src)!);
    return;
  }

  const prefix = src === '/' ? '/' : `${src}/`;
  ensureSnapshotParents(snapshot, parentOf(dst));
  snapshot.dirs.add(dst);

  for (const dirPath of [...snapshot.dirs]) {
    if (dirPath !== src && dirPath.startsWith(prefix)) {
      snapshot.dirs.add(dst + dirPath.slice(src.length));
    }
  }

  for (const [filePath, size] of snapshot.files.entries()) {
    if (filePath.startsWith(prefix)) {
      snapshot.files.set(dst + filePath.slice(src.length), size);
    }
  }
}

export function applyMoveToSnapshot(snapshot: VfsSnapshot, src: string, dst: string): void {
  if (src === dst) {
    return;
  }
  applyCopyToSnapshot(snapshot, src, dst);
  applyDeleteToSnapshot(snapshot, src);
}

export function enforceResourcePolicy(
  snapshot: VfsSnapshot,
  policy: ResolvedVfsResourcePolicy,
  contextPath: string,
): void {
  if (!hasResourcePolicy(policy)) {
    return;
  }

  let totalBytes = 0;
  const childrenByDir = new Map<string, number>();

  for (const dirPath of snapshot.dirs) {
    if (dirPath !== '/') {
      const depth = pathDepth(dirPath);
      if (policy.maxDirectoryDepth !== undefined && depth > policy.maxDirectoryDepth) {
        throw limitError(contextPath, 'maxDirectoryDepth', policy.maxDirectoryDepth, depth);
      }
      incrementChildCount(childrenByDir, parentOf(dirPath));
    }
  }

  for (const [filePath, size] of snapshot.files) {
    totalBytes += size;

    if (policy.maxFileBytes !== undefined && size > policy.maxFileBytes) {
      throw limitError(filePath, 'maxFileBytes', policy.maxFileBytes, size);
    }
    if (
      policy.maxOutputArtifactBytes !== undefined &&
      isRuntimeArtifactPath(filePath) &&
      size > policy.maxOutputArtifactBytes
    ) {
      throw limitError(filePath, 'maxOutputArtifactBytes', policy.maxOutputArtifactBytes, size);
    }

    incrementChildCount(childrenByDir, parentOf(filePath));
  }

  if (policy.maxTotalBytes !== undefined && totalBytes > policy.maxTotalBytes) {
    throw limitError(contextPath, 'maxTotalBytes', policy.maxTotalBytes, totalBytes);
  }

  if (policy.maxEntriesPerDirectory !== undefined) {
    for (const [dirPath, count] of childrenByDir) {
      if (count > policy.maxEntriesPerDirectory) {
        throw limitError(dirPath, 'maxEntriesPerDirectory', policy.maxEntriesPerDirectory, count);
      }
    }
  }
}

export function createSnapshotFromEntries(
  files: Iterable<[string, number]>,
  dirs: Iterable<string>,
): VfsSnapshot {
  return {
    files: new Map(files),
    dirs: new Set(dirs),
  };
}

export function isResourceLimitError(value: unknown): boolean {
  return (
    !!value &&
    typeof value === 'object' &&
    'code' in value &&
    (value as { code?: unknown }).code === 'ERESOURCE_LIMIT'
  );
}

function ensureSnapshotParents(snapshot: VfsSnapshot, path: string): void {
  if (path === '/') {
    snapshot.dirs.add('/');
    return;
  }

  const segments = path.split('/').filter(Boolean);
  let current = '';
  for (const segment of segments) {
    current += `/${segment}`;
    snapshot.dirs.add(current);
  }
}

function incrementChildCount(childrenByDir: Map<string, number>, dirPath: string): void {
  childrenByDir.set(dirPath, (childrenByDir.get(dirPath) ?? 0) + 1);
}

function pathDepth(path: string): number {
  return path.split('/').filter(Boolean).length;
}

function isRuntimeArtifactPath(path: string): boolean {
  return path.startsWith('/.system/cmd-output/');
}

function limitError(path: string, limitKind: keyof VfsResourcePolicy, limit: number, actual: number) {
  return vfsError('ERESOURCE_LIMIT', path, {
    details: {
      limitKind,
      limit,
      actual,
    },
  });
}

function validateLimit(name: string, value: number | undefined): void {
  if (value === undefined) {
    return;
  }
  if (!Number.isInteger(value) || value < 0) {
    throw new Error(`${name} must be a non-negative integer`);
  }
}

File: python/src/onetool/vfs/policy.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable

from .errors import to_vfs_error, vfs_error
from .path_utils import parent_of


@dataclass(frozen=True, slots=True)
class VfsResourcePolicy:
    max_file_bytes: int | None = None
    max_total_bytes: int | None = None
    max_directory_depth: int | None = None
    max_entries_per_directory: int | None = None
    max_output_artifact_bytes: int | None = None


ResolvedVfsResourcePolicy = VfsResourcePolicy


@dataclass(slots=True)
class VfsSnapshot:
    files: dict[str, int]
    dirs: set[str]


EMPTY_RESOURCE_POLICY = VfsResourcePolicy()


def resolve_vfs_resource_policy(policy: VfsResourcePolicy | None = None) -> ResolvedVfsResourcePolicy:
    if policy is None:
        return EMPTY_RESOURCE_POLICY

    _validate_limit("maxFileBytes", policy.max_file_bytes)
    _validate_limit("maxTotalBytes", policy.max_total_bytes)
    _validate_limit("maxDirectoryDepth", policy.max_directory_depth)
    _validate_limit("maxEntriesPerDirectory", policy.max_entries_per_directory)
    _validate_limit("maxOutputArtifactBytes", policy.max_output_artifact_bytes)

    return VfsResourcePolicy(
        max_file_bytes=policy.max_file_bytes,
        max_total_bytes=policy.max_total_bytes,
        max_directory_depth=policy.max_directory_depth,
        max_entries_per_directory=policy.max_entries_per_directory,
        max_output_artifact_bytes=policy.max_output_artifact_bytes,
    )


def has_resource_policy(policy: ResolvedVfsResourcePolicy) -> bool:
    return any(
        value is not None
        for value in (
            policy.max_file_bytes,
            policy.max_total_bytes,
            policy.max_directory_depth,
            policy.max_entries_per_directory,
            policy.max_output_artifact_bytes,
        )
    )


def clone_snapshot(snapshot: VfsSnapshot) -> VfsSnapshot:
    return VfsSnapshot(files=dict(snapshot.files), dirs=set(snapshot.dirs))


def create_empty_snapshot() -> VfsSnapshot:
    return VfsSnapshot(files={}, dirs={"/"})


def create_snapshot_from_entries(files: Iterable[tuple[str, int]], dirs: Iterable[str]) -> VfsSnapshot:
    return VfsSnapshot(files=dict(files), dirs=set(dirs))


def apply_mkdir_to_snapshot(snapshot: VfsSnapshot, path: str, parents: bool) -> None:
    if path == "/":
        return
    if parents:
        _ensure_snapshot_parents(snapshot, path)
        return
    snapshot.dirs.add(path)


def apply_write_to_snapshot(snapshot: VfsSnapshot, path: str, size: int, make_parents: bool) -> None:
    if make_parents:
        _ensure_snapshot_parents(snapshot, parent_of(path))
    snapshot.files[path] = size


def apply_append_to_snapshot(snapshot: VfsSnapshot, path: str, size_delta: int, make_parents: bool) -> None:
    if make_parents:
        _ensure_snapshot_parents(snapshot, parent_of(path))
    snapshot.files[path] = snapshot.files.get(path, 0) + size_delta


def apply_delete_to_snapshot(snapshot: VfsSnapshot, path: str) -> None:
    if path in snapshot.files:
        del snapshot.files[path]
        return

    prefix = f"{path}/"
    for file_path in list(snapshot.files):
        if file_path.startswith(prefix):
            del snapshot.files[file_path]

    for dir_path in list(snapshot.dirs):
        if dir_path == path or dir_path.startswith(prefix):
            snapshot.dirs.remove(dir_path)

    snapshot.dirs.add("/")


def apply_copy_to_snapshot(snapshot: VfsSnapshot, src: str, dst: str) -> None:
    if src in snapshot.files:
        _ensure_snapshot_parents(snapshot, parent_of(dst))
        snapshot.files[dst] = snapshot.files[src]
        return

    prefix = "/" if src == "/" else f"{src}/"
    _ensure_snapshot_parents(snapshot, parent_of(dst))
    snapshot.dirs.add(dst)

    for dir_path in list(snapshot.dirs):
        if dir_path != src and dir_path.startswith(prefix):
            snapshot.dirs.add(dst + dir_path[len(src) :])

    for file_path, size in list(snapshot.files.items()):
        if file_path.startswith(prefix):
            snapshot.files[dst + file_path[len(src) :]] = size


def apply_move_to_snapshot(snapshot: VfsSnapshot, src: str, dst: str) -> None:
    if src == dst:
        return
    apply_copy_to_snapshot(snapshot, src, dst)
    apply_delete_to_snapshot(snapshot, src)


def enforce_resource_policy(
    snapshot: VfsSnapshot,
    policy: ResolvedVfsResourcePolicy,
    context_path: str,
) -> None:
    if not has_resource_policy(policy):
        return

    total_bytes = 0
    children_by_dir: dict[str, int] = {}

    for dir_path in snapshot.dirs:
        if dir_path != "/":
            depth = _path_depth(dir_path)
            if policy.max_directory_depth is not None and depth > policy.max_directory_depth:
                raise _limit_error(context_path, "maxDirectoryDepth", policy.max_directory_depth, depth)
            _increment_child_count(children_by_dir, parent_of(dir_path))

    for file_path, size in snapshot.files.items():
        total_bytes += size

        if policy.max_file_bytes is not None and size > policy.max_file_bytes:
            raise _limit_error(file_path, "maxFileBytes", policy.max_file_bytes, size)

        if (
            policy.max_output_artifact_bytes is not None
            and _is_runtime_artifact_path(file_path)
            and size > policy.max_output_artifact_bytes
        ):
            raise _limit_error(
                file_path,
                "maxOutputArtifactBytes",
                policy.max_output_artifact_bytes,
                size,
            )

        _increment_child_count(children_by_dir, parent_of(file_path))

    if policy.max_total_bytes is not None and total_bytes > policy.max_total_bytes:
        raise _limit_error(context_path, "maxTotalBytes", policy.max_total_bytes, total_bytes)

    if policy.max_entries_per_directory is not None:
        for dir_path, count in children_by_dir.items():
            if count > policy.max_entries_per_directory:
                raise _limit_error(
                    dir_path,
                    "maxEntriesPerDirectory",
                    policy.max_entries_per_directory,
                    count,
                )


def check_resource_limit_error(value: object) -> bool:
    caught = to_vfs_error(value)
    return caught is not None and caught.code == "ERESOURCE_LIMIT"


def _ensure_snapshot_parents(snapshot: VfsSnapshot, path: str) -> None:
    if path == "/":
        snapshot.dirs.add("/")
        return

    current = ""
    for segment in filter(None, path.split("/")):
        current += f"/{segment}"
        snapshot.dirs.add(current)


def _increment_child_count(children_by_dir: dict[str, int], dir_path: str) -> None:
    children_by_dir[dir_path] = children_by_dir.get(dir_path, 0) + 1


def _path_depth(path: str) -> int:
    return len([segment for segment in path.split("/") if segment])


def _is_runtime_artifact_path(path: str) -> bool:
    return path.startswith("/.system/cmd-output/")


def _limit_error(path: str, limit_kind: str, limit: int, actual: int) -> VfsError:
    return vfs_error(
        "ERESOURCE_LIMIT",
        path,
        details={
            "limitKind": limit_kind,
            "limit": limit,
            "actual": actual,
        },
    )


def _validate_limit(name: str, value: int | None) -> None:
    if value is None:
        return
    if not isinstance(value, int) or value < 0:
        raise ValueError(f"{name} must be a non-negative integer")
